### PR TITLE
feat(backend): add oauth2 for all endpoints #75

### LINKED ISSUE #75: Notification backend: add authentication
Integrate the backend with the GitHub login. For now only check that the user is logged in.

### PR DESCRIPTION
**Pull request by JonasKellerer**
resolves #75

### Summary
This adds outh2 authentification to the backend. We use the same cookies we produce on the website. It is currently not possible to add the auth to swagger. This will be done in #116

### CODE REVIEW

**Review by fengelniederhammer — CHANGES_REQUESTED** [review]
How can I test it? What do I need to send to the API?

File: backend/src/test/kotlin/org/genspectrum/dashboardsbackend/controller/TestUser.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.genspectrum.dashboardsbackend.controller
+
+import org.springframework.security.core.authority.SimpleGrantedAuthority
+import org.springframework.security.oauth2.core.user.DefaultOAuth2User
+
+fun getLoggedInUser() = oauthUser
+
+val oauthUser = DefaultOAuth2User(
```
[the remark this answers is not in the document]
Comment (JonasKellerer) [reply backend/src/test/kotlin/org/genspectrum/dashboardsbackend/controller/TestUser.kt:8]: This is only the starting point for creating special users with special properties, e.g. name. They are not used now, but they will be in #76.

File: backend/src/test/kotlin/org/genspectrum/dashboardsbackend/controller/SubscriptionsClient.kt [hunk 2]
```diff
@@ -16,31 +19,66 @@ class SubscriptionsClient(
...
-        getSubscriptionRaw(id)
-            .andExpect(status().isOk),
+    private fun postWithUser(url: String, content: String, user: OAuth2User) = mockMvc.perform(
+        post(url)
+            .content(content)
+            .contentType(MediaType.APPLICATION_JSON)
+            .with(
+                oauth2Login().oauth2User(user),
+            ).with(csrf()),
     )
```
[the remark this answers is not in the document]
Comment (JonasKellerer) [reply backend/src/test/kotlin/org/genspectrum/dashboardsbackend/controller/SubscriptionsClient.kt]: Nice idea. I never used extension functions.

File: backend/src/test/resources/application.properties [hunk 3]
```diff
@@ -2,3 +2,6 @@ spring.datasource.driverClassName=org.testcontainers.jdbc.ContainerDatabaseDrive
 spring.datasource.url=jdbc:tc:postgresql:latest:///subscriptions
 spring.datasource.username=test
 spring.datasource.password=test
+
+spring.security.oauth2.client.registration.github.client-id=test-client-id
+spring.security.oauth2.client.registration.github.client-secret=test-client-secret
```
[the remark this answers is not in the document]
Comment (JonasKellerer) [reply backend/src/test/resources/application.properties:7]: Yeah. My local configuration had it.

**Comment by JonasKellerer** [comment]
This is now a WIP to show the general idea behind the backend proxy with astro.

**Comment by JonasKellerer** [comment]
This is now a proper PR again:

## Summary
We added a proxy for the spring backend, by having the client route `/api/subscriptions`. This way users need to be logged in to access the spring backend. The spring backend itself will be protected, because it is not exposed to the internet. Only the astro backend will call the spring backend.

File: website/src/config.ts [hunk 4]
```diff
@@ -29,6 +29,11 @@ const dashboardsConfigSchema = z
                 clientId: z.string(),
             }),
         }),
+    })
+    .extend({
```
Comment (fengelniederhammer) [inline website/src/config.ts:33]: We could simply merge this we the `extend` above.
```suggestion
```

File: website/src/pages/api/subscriptions/[...path].ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        });
+    } catch (error) {
+        return getInternalErrorResponse(request.url);
+    }
+}
+
+function getBackendUrl(request: Request, session: Session) {
+    const backendEndpoint = new URL(request.url).pathname.slice(API_PATHNAME_LENGTH);
+    const backendUrl = new URL(backendEndpoint, getDashboardsConfig().backend.url);
+    backendUrl.searchParams.set('userId', session.user?.id || '');
```
Comment (fengelniederhammer) [inline website/src/pages/api/subscriptions/[...path].ts:33]: We already check in the calling function that the userId is set, so we can get rid of this confusing check:

```suggestion
function getBackendUrl(request: Request, userId: string) {
    const backendEndpoint = new URL(request.url).pathname.slice(API_PATHNAME_LENGTH);
    const backendUrl = new URL(backendEndpoint, getDashboardsConfig().backend.url);
    backendUrl.searchParams.set('userId', userId);
```

File: website/src/pages/api/subscriptions/[...path].ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        headers: {
+            'Content-Type': 'application/json',
+        },
+    });
+};
+
+const getInternalErrorResponse = (requestUrl: string) => {
+    const response: BackendError = {
+        title: 'Internal Server Error',
+        detail: "Couldn't fetch data from the backend",
```
Comment (fengelniederhammer) [inline website/src/pages/api/subscriptions/[...path].ts:63]: Does the `error` contain more detail about why it fails? It would be good to at least log it somehow, or even write it to the response.

File: website/src/env.d.ts [hunk 7]
```diff
@@ -1 +1,3 @@
+/// <reference path="../.astro/types.d.ts" />
 /// <reference types="astro/client" />
+declare module 'set-cookie-parser';
```
Comment (fengelniederhammer) [inline website/src/env.d.ts:3]: Why is this necessary?

File: website/src/pages/api/subscriptions/[...path].ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    if (!session || !session.user?.id) {
+        return getUnauthorizedResponse(request.url);
+    }
+
+    const backendUrl = getBackendUrl(request, session);
+
+    try {
+        const response = await fetch(backendUrl, {
+            method: request.method,
+        });
```
Comment (fengelniederhammer) [inline website/src/pages/api/subscriptions/[...path].ts:19]: We also need to forward the response body.

File: website/src/pages/api/subscriptions/[...path].ts (hunk 5 above)
Comment (fengelniederhammer) [inline website/src/pages/api/subscriptions/[...path].ts:33]: We also need to pass on the search params of the original request.

File: website/src/pages/api/subscriptions/[...path].ts
Comment (fengelniederhammer) [inline website/src/pages/api/subscriptions/[...path].ts:1]: I think it would be good to add a unit test with msw for this file to make sure that the forwarding works as expected. We can also do that later, when we're sure that we'll settle for this solution.

**Review by fengelniederhammer — CHANGES_REQUESTED** [review]

File: website/src/env.d.ts (hunk 7 above)
Comment (JonasKellerer) [reply website/src/env.d.ts:3]: When I don't add this I get this error from `tsc`:

```
29 import { parseString } from 'set-cookie-parser'
                               ~~~~~~~~~~~~~~~~~~~
Found 1 error in node_modules/auth-astro/server.ts:29
```
This is inside of one of our imported libraries (auth-astro), and I did not find any other way, that typescript does not check our dependencies as well... Do you know of a way?

Comment (fengelniederhammer) [reply website/src/env.d.ts:3]: Fine by me. Probably also doesn't hurt. Maybe add a comment on why the `declare` is necessary?

File: website/src/pages/api/subscriptions/[...path].ts (hunk 6 above)
Comment (JonasKellerer) [reply website/src/pages/api/subscriptions/[...path].ts:63]: Currently, this is only the fetch failed error. The other errors will be forwarded from the backend directly, so does it make sense to add the detail here?

Comment (fengelniederhammer) [reply website/src/pages/api/subscriptions/[...path].ts:63]: Maybe a `console.error(error)` on the calling side of this function already does the job.
(We should think of a logging solution for the website...)

File: website/src/pages/api/subscriptions/[...path].ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        return getUnauthorizedResponse(request.url);
+    }
+
+    const backendUrl = getBackendUrl(request, session.user.id);
+
+    try {
+        const response = await fetch(backendUrl, {
+            method: request.method,
+            body: request.body,
+        });
```
Comment (fengelniederhammer) [inline website/src/pages/api/subscriptions/[...path].ts:19]: POST requests don't work yet. I think we can simply do this:
```suggestion
        const response = await fetch(backendUrl, request);
```

**Review by fengelniederhammer — APPROVED** [review]

File: website/src/pages/api/subscriptions/[...path].ts (hunk 9 above)
Comment (JonasKellerer) [reply website/src/pages/api/subscriptions/[...path].ts]: POST works now
